tests/e2e/ctl_v3_auth_cluster_test.go

```diff
@@ -0,0 +1,125 @@
+// Copyright 2022 The etcd Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package e2e
+
+import (
+	"context"
+	"fmt"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+
+	clientv3 "go.etcd.io/etcd/client/v3"
+	"go.etcd.io/etcd/tests/v3/framework/config"
+	"go.etcd.io/etcd/tests/v3/framework/e2e"
+)
+
+func TestAuthCluster(t *testing.T) {
+	e2e.BeforeTest(t)
+	cfg := &e2e.EtcdProcessClusterConfig{
+		ClusterSize:   1,
+		InitialToken:  "new",
+		SnapshotCount: 2,
+	}
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	epc, err := e2e.NewEtcdProcessCluster(ctx, t, cfg)
+	if err != nil {
+		t.Fatalf("could not start etcd process cluster (%v)", err)
+	}
+	defer func() {
+		if err := epc.Close(); err != nil {
+			t.Fatalf("could not close test cluster (%v)", err)
+		}
+	}()
+
+	epcClient := epc.Client()
+	createUsers(ctx, t, epcClient)
+
+	if err := epcClient.AuthEnable(ctx); err != nil {
+		t.Fatalf("could not enable Auth: (%v)", err)
+	}
+
+	testUserClientOpts := e2e.WithAuth("test", "testPassword")
+	rootUserClientOpts := e2e.WithAuth("root", "rootPassword")
+
+	// write more than SnapshotCount keys to single leader to make sure snapshot is created
+	for i := 0; i <= 10; i++ {
+		if err := epc.Client(testUserClientOpts).Put(ctx, fmt.Sprintf("/test/%d", i), "test", config.PutOptions{}); err != nil {
+			t.Fatalf("failed to Put (%v)", err)
+		}
+	}
+
+	// start second process
+	if err := epc.StartNewProc(ctx, t, rootUserClientOpts); err != nil {
+		t.Fatalf("could not start second etcd process (%v)", err)
+	}
```

Comment:
This won't trigger transfer db snapshot to new member and restore auth store from backend. 

By default `SnapshotCatchUpEntries` is 5k. 

https://github.com/etcd-io/etcd/blob/0dfd7264c0b7c8e02e4a7c45259d35543a02d462/server/config/config.go#L53-L59

So I doubt this test really protects the issue from happening again while the fix makes sense to me. @veshij @ahrtr

Reply:
Good catch! The reason why we can successfully & manually reproduce & verify this is that the snapi is usually < 5K, it always compacts all entries starting from index 2. https://github.com/etcd-io/etcd/blob/0dfd7264c0b7c8e02e4a7c45259d35543a02d462/server/etcdserver/server.go#L2099-L2102

Two points:
1. This might not be correct, we shouldn't compact the log at all if snapi < 5K. Of course, it isn't a big deal. 
2. In an e2e or integration test case, the snapi should be usually < 5K, so the test case should be still valid. But I agree that we should correct it. So let's write more than 5K K/V to make it correct and stable. 

@chaochn47 @veshij Please feel free to deliver a PR to address it for both `main` and `release-3.5`. Let fix # 2 firstly, and probably @ 1 separately.

Reply:
Just to confirm - we need to only increase number of keys written to 5K (>= SnapshotCatchUpEntries)?

Reply:
YES

Reply:
> The reason why we can successfully & manually reproduce & verify this is that the snapi is usually < 5K, it always compacts all entries starting from index 2. 

I am a little bit confused about this statement. How this test reproduced the scenario that db snapshot is transferred to new member? @ahrtr 

Looks like raft log won't be compacted unless the very first one entry. 
https://github.com/etcd-io/etcd/blob/0dfd7264c0b7c8e02e4a7c45259d35543a02d462/raft/storage.go#L227-L248

Reply:
I obviously don't fully understand the replication behavior here, but what I noticed while writing the test - without SnapshotCount set to some low value I wasn't able to trigger the issue. Based on logs it looked like raft log was applied instead of snapshot.

Reply:
Correction to my previous typo: all entries starting from index 2 will be retained.  Since the first entry is compacted, so when a new member is added, the leader will still send a snapshot to the new member. 

So let's add the following two cases:
1. keep the current case;
2. write more than 5K K/V, e.g 5100, and keep all others unchanged.

>  This won't trigger transfering db snapshot to new member and restore auth store from backend. Confirmed by local testing.

I am curious how did you confirm that?

Reply:
Thanks for the explanation!! Just verified again new member will always receive a snapshot from leader whenever the configured snapshot count < KV writes. 

```
/home/chaochn/go/src/go.etcd.io/etcd/bin/etcd (TestAuthCluster-test-0) (23533): 
{
	"level": "info",
	"ts": "2022-11-04T06:05:50.932Z",
	"caller": "etcdserver/snapshot_merge.go:66",
	"msg": "sent database snapshot to writer",
	"bytes": 20480,
	"size": "20 kB"
}
/home/chaochn/go/src/go.etcd.io/etcd/bin/etcd (TestAuthCluster-test-1) (23773): 
{
	"level": "info",
	"ts": "2022-11-04T06:05:50.932Z",
	"caller": "rafthttp/http.go:257",
	"msg": "receiving database snapshot",
	"local-member-id": "4e6825c690b0cb86",
	"remote-snapshot-sender-id": "ca50e9357181d758",
	"incoming-snapshot-index": 48,
	"incoming-snapshot-message-size-bytes": 7723,
	"incoming-snapshot-message-size": "7.7 kB"
}
```

I must be confused by existing member restart and new member joining the cluster.

The former requires at least 5k KV writes to trigger a db snapshot.